### PR TITLE
MacOS build script & Github Action for Trelis 17.1 and Cubit 2020.2

### PR DESCRIPTION
**Pull request by bam241**
This builds on top of #84 , review after merge.
This also relies on svalinn/mcnp2cad#73

This includes:
- build script to build the plugin on MacOS
- GItHub Action to build the plugin on MacOS for Trelis17.1 and Cubit 2020.2

### CODE REVIEW

File: .github/workflows/mac.yml
Comment (bam241) [inline .github/workflows/mac.yml]: this needs to be changed when svalinn/mcnp2cad#73, gets merged.

File: .github/workflows/mac.yml
Comment (bam241) [inline .github/workflows/mac.yml]: this needs to be changed when svalinn/mcnp2cad#73, gets merged.

**Comment by bam241** [comment]
this works hdf5@1.12 added a new PR depedency: svalinn/dagmc#728

we could avoid svalinn/dagmc#728 and limit hdf5 version to `1.10.7`

**Comment by bam241** [comment]
getting the following error when trying to load Cubit2020.2 plugin built with the GitHub action:
```
ERROR: <stdin>, line 0. Error loading /Applications/Coreform-Cubit-2020.2.app/Contents/MacOS/plugins/libsvalinn_plugin.so:
ERROR: <stdin>, line 0. 
Finished Command: undo on
```

**Comment by bam241** [comment]
installing `brew install eigen hdf5` using brew solves it.

I guess we have to ship some of those within the package

**Comment by bam241** [comment]
it was actually slightly more complicated than that:
one of the `hdf5` deps was required not` hdf5` itself `brew install szip`

**Comment by bam241** [comment]
I fixed it adding `/libsz.2.dylib` in the plugin and running :
`sudo install_name_tool -change /usr/local/opt/szip/lib/libsz.2.dylib @rpath//libsz.2.dylib libsvalinn_plugin.so`

**Comment by bam241** [comment]
inclusion of szlib fixed it.

**Comment by gonuke** [comment]
I plan to review this after we merge #84 to focus on the Mac parts...

**Comment by bam241** [comment]
Reworking this so it can match the linux design, 
It'll be broken for a while, please ignore this until I ping you :)

**Comment by bam241** [comment]
@gonuke @pshriwise I reshaped the MacOS build to leverage the linux_share_build.sh (now named unix_shared_build.sh)
I have no way right now to test the `build_plugin_macos.sh`.

I didn't re-tried the linux docker build against those new change as well.

File: .github/workflows/mac.yml
Comment (gonuke) [inline .github/workflows/mac.yml]: It is weird that this doesn't get sourced automatically...  I wonder if it is sourcing something else? (e.g. possibly .bash_profile per bash documentation?)  Alternatively, is there a way we can launch bash in the `shell` command to ensure it is sourced?

File: .github/workflows/mac.yml
Comment (gonuke) [inline .github/workflows/mac.yml:1]: In general, I wonder if we can nudge this to be even more similar to the linux action, just for long term maintenance purposes

File: scripts/CubitExport-release.cmake
Comment (gonuke) [inline scripts/CubitExport-release.cmake]: Do we need this file? or is it part of the SDK?

File: scripts/CubitGeomConfig.cmake
Comment (gonuke) [inline scripts/CubitGeomConfig.cmake]: Do we need this file in the repo?

File: scripts/build_plugin_linux.sh
Comment (gonuke) [inline scripts/build_plugin_linux.sh]: Should probably leave the reference to Trelis here as an option since it 17.1.0 is still called Trelis.
```suggestion
# $1 is the version of Trelis/Cubit one are trying to compile against i.e. 17.1.0
```

File: scripts/build_plugin_linux.sh
Comment (gonuke) [inline scripts/build_plugin_linux.sh]: It looks like we're largely changing things to refer to Cubit instead of Trelis - should we change this function name, as well?

File: scripts/build_plugin_macos.sh
Comment (gonuke) [inline scripts/build_plugin_macos.sh]: This doesn't look right...

File: scripts/build_plugin_macos.sh
Comment (gonuke) [inline scripts/build_plugin_macos.sh]: This doesn't look like its using anything out of `unix_share_build.sh` yet?

**Review by gonuke — CHANGES_REQUESTED** [review]
Maybe your comment above answers one of my questions - but it doesn't look like `build_plugin_macos.sh` uses any of the `unix_shared_build.sh` yet...

File: .github/workflows/mac.yml
Comment (bam241) [reply .github/workflows/mac.yml]: sadly the bash shell is run internally with the --no-profile argument which prevent on macOS to source profile/bash_profile by default :(
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#using-a-specific-shell
we don't have the same behavior with linux...

File: scripts/CubitExport-release.cmake
Comment (bam241) [reply scripts/CubitExport-release.cmake]: yes it is part of the sdk, I had to tweak it to have it work on Mac, 
some paths were not provided correctly...

File: scripts/build_plugin_linux.sh
Comment (bam241) [reply scripts/build_plugin_linux.sh]: I think this is the way forward...

File: scripts/build_plugin_macos.sh
Comment (bam241) [reply scripts/build_plugin_macos.sh]: no I did not touch this one yet, as I don't have access to the MACOS vm.

File: scripts/CubitGeomConfig.cmake
Comment (bam241) [reply scripts/CubitGeomConfig.cmake]: I don't quite remember precisely, but it was part of the file I had to modify to have the plugin make to work.

I can do some test to check if this is still necessary...

File: scripts/build_plugin_linux.sh
Comment (bam241) [reply scripts/build_plugin_linux.sh]: I did a Trelis to CUBIT replacement for ENV variable, it caught to much stuff... need to revert some of them...

File: .github/workflows/mac.yml
Comment (gonuke) [reply .github/workflows/mac.yml]: odd - it says that Linux should behave the same....

File: scripts/CubitExport-release.cmake
Comment (gonuke) [reply scripts/CubitExport-release.cmake]: I wonder if we should provide that differently then???  It seems to make more sense to patch the SDK that include the whole file here, but I'm not sure the best way to do that....

**Comment by bam241** [comment]
our PyNE amalgamation broke this...

File: scripts/CubitExport-release.cmake
Comment (bam241) [reply scripts/CubitExport-release.cmake]: we could get a diff, and apply it ?

File: .github/workflows/mac.yml
Comment (gonuke) [reply .github/workflows/mac.yml]: Actually... our declaration of the shell should override this on both platforms

I wonder if there is also a `.bash_profile` that is being read instead when we use `-l`?

File: scripts/build_plugin_macos.sh
Comment (gonuke) [reply scripts/build_plugin_macos.sh]: So what's the next step for this file?

Comment (bam241) [reply scripts/build_plugin_macos.sh]: maybe I can remove it from this. 
and PR it when I am back ?

File: .github/workflows/mac.yml
Comment (gonuke) [inline .github/workflows/mac.yml]: Should these be changed to CUBIT as well?
```suggestion
            echo "CUBIT_PKG=Trelis-17.1.0-Mac64.dmg" >> $GITHUB_ENV
            echo "CUBIT_SDK_PKG=Trelis-SDK-17.1.0-Mac64.tar.gz" >> $GITHUB_ENV
```

**Review by gonuke — CHANGES_REQUESTED** [review]
Some lingering references to trelis in variable names

**Comment by gonuke** [comment]
Still curious about making the `mac.yml` file a little more similar to the `linux.yml` file, but that might be best as a conversation

**Comment by gonuke** [comment]
I PR'ed some changes to address

**Review by gonuke — APPROVED** [review]
Given my contribution, I probably shouldn't merge this.... but I approve

**Comment by gonuke** [comment]
I think this is ready to go @pshriwise

**Comment by bam241** [comment]
as a reminder this depends on svalinn/mcnp2cad#73

**Comment by bam241** [comment]
@pshriwise when you have time could you take a look here ?

File: .github/workflows/mac.yml
Comment (pshriwise) [reply .github/workflows/mac.yml:1]: Seconded. I wonder if a  job parameter indicating the OS type and some [conditionals](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsif) in the job steps that check that parameter 
would do the trick.

File: scripts/CubitExport-release.cmake
Comment (pshriwise) [reply scripts/CubitExport-release.cmake]: That sounds like a good plan to me @bam241!

File: scripts/CubitGeomConfig.cmake
Comment (pshriwise) [reply scripts/CubitGeomConfig.cmake]: Maybe we can add this to the list of files to get diffs of for a patch.

File: scripts/unix_share_build.sh
Comment (pshriwise) [inline scripts/unix_share_build.sh:75]: I'm curious about this. Why did it need to change? Should we select only a few processes instead?

**Review by pshriwise — CHANGES_REQUESTED** [review]
Overall this looks pretty good to me. There are just some outstanding issues that @gonuke raised I think:

  - looking into merging the Linux and Mac workflow files
  - getting diff's of the Cubit CMake files from the SDK and look into patching those instead of including the modified files

Also if we can communicate the changes we needed to make to the CMake files Coreform that would probably be good.

File: scripts/unix_share_build.sh
Comment (bam241) [reply scripts/unix_share_build.sh:75]: I removed it because we are not here for speed, and this something causes problems with `gcc`....

File: scripts/build_plugin_linux.sh
Comment (bam241) [inline scripts/build_plugin_linux.sh]:
```suggestion
linux_install_prerequisites
```

**Comment by bam241** [comment]
@pshriwise @gonuke I replaced the cmake files with a .diff files and apply a patch now instead of replacing those 2 files...

I think this is ready to go in
(as we said the merge between Mac.yml and linux.yml should go in a later PR)

File: scripts/build_plugin_linux.sh
Comment (bam241) [inline scripts/build_plugin_linux.sh]:
```suggestion
linux_build_plugin_pkg $1
```

File: scripts/unix_share_build.sh
Comment (pshriwise) [inline scripts/unix_share_build.sh:153]: Should these be `$SUDO` instead of `sudo`? I can't recall.

File: scripts/unix_share_build.sh
Comment (pshriwise) [inline scripts/unix_share_build.sh:137]: Gonna request a comment or two here on the different setup methods and commands used.

File: CMakeLists.txt
Comment (pshriwise) [inline CMakeLists.txt:51]: Does use of HDF5 static libraries have any implications for how we build our DAGMC or plugin libraries?

File: scripts/unix_share_build.sh
Comment (pshriwise) [inline scripts/unix_share_build.sh:174]: Some comments for what we're doing here would be nice also.

File: scripts/unix_share_build.sh
Comment (bam241) [reply scripts/unix_share_build.sh:153]: I used `sudo` for Mac instead of `$SUDO` because it should be required regardless of the platform you are using (docker, CI, local) where on ubuntu `sudo` is not required on docker as we run as root...

File: CMakeLists.txt
Comment (bam241) [reply CMakeLists.txt:51]: one need to use static HDF5 to allow o be build the plugin as a single `.dylib`

Or at least this is what I understood from my digging few months ago....

File: scripts/unix_share_build.sh
Comment (pshriwise) [reply scripts/unix_share_build.sh:153]: Ah, nice. Didn't know that you don't get root access in an OSX docker image.

**Review by pshriwise — APPROVED** [review]
Thanks @bam241! On to the next PR I suppose 😃

**Pull request merged**
